dotnet/src/SemanticKernel.Core/KernelBuilder.cs

```diff
@@ -1,50 +1,87 @@
-    public static Kernel Create()
-    {
-        var builder = new KernelBuilder();
-        return builder.Build();
-    }
+    /// <summary>Initializes a new instance of the <see cref="KernelBuilder"/>.</summary>
+    public KernelBuilder() { }
 
-    /// <summary>
-    /// Build a new kernel instance using the settings passed so far.
-    /// </summary>
-    /// <returns>Kernel instance</returns>
+    /// <summary>Constructs a new instance of <see cref="Kernel"/> using all of the settings configured on the builder.</summary>
+    /// <returns>The new <see cref="Kernel"/> instance.</returns>
+    /// <remarks>
+    /// Every call to <see cref="Build"/> produces a new <see cref="Kernel"/> instance. The resulting <see cref="Kernel"/>
+    /// instances will not share the same plugins collection or services provider (unless there are no services).
+    /// </remarks>
     public Kernel Build()
     {
-#pragma warning disable CS8604 // Possible null reference argument.
-        var instance = new Kernel(
-            this._aiServices.Build(),
-            new KernelPluginCollection(),
-            this._serviceSelector,
-            this._httpHandlerFactory,
-            this._loggerFactory
-        );
-#pragma warning restore CS8604 // Possible null reference argument.
+        IServiceProvider serviceProvider;
+        if (this._services is { Count: > 0 } services)
+        {
+            // This is a workaround for Microsoft.Extensions.DependencyInjection's GetKeyedServices not currently supporting
+            // enumerating all services for a given type regardless of key.
+            // https://github.com/dotnet/runtime/issues/91466
+            // We need this support to, for example, allow IServiceSelector to pick from multiple named instances of an AI
+            // service based on their characteristics. Until that is addressed, we work around it by injecting as a service all
+            // of the keys used for a given type, such that Kernel can then query for this dictionary and enumerate it. This means
+            // that such functionality will work when KernelBuilder is used to build the kernel but not when the IServiceProvider
+            // is created via other means, such as if Kernel is directly created by DI. However, it allows us to create the APIs
+            // the way we want them for the longer term and then subsequently fix the implementation when M.E.DI is fixed.
+            Dictionary<Type, List<object?>> typeToKeyMappings = new();
+            foreach (ServiceDescriptor serviceDescriptor in services)
+            {
+                if (!typeToKeyMappings.TryGetValue(serviceDescriptor.ServiceType, out List<object?>? keys))
+                {
+                    typeToKeyMappings[serviceDescriptor.ServiceType] = keys = new();
+                }
+
+                keys.Add(serviceDescriptor.ServiceKey);
+            }
+            services.AddKeyedSingleton(Kernel.ServiceTypeToKeyMappingsKey, typeToKeyMappings);
```

Comment:
Could we do this every time `Kernel.GetAllServices` is called?

Reply:
Why?

Reply:
Answered above